File: examples/agent_harness.ts
/**
 * agent_harness.ts — A lightweight, async-native, single-file LLM agent harness.
 *
 * Built on the OpenAI SDK (compatible with litellm proxy, OpenAI, Anthropic via proxy, etc.)
 * Features:
 *   - Decorator-free tool registration with auto-schema from zod or manual schemas
 *   - Async middleware pipeline (pre/post processing of messages and responses)
 *   - Async hook system for lifecycle events
 *   - Streaming support
 *   - Parallel tool execution
 *   - Configurable retry with exponential backoff
 *
 * Usage:
 *   import { Agent, defineTool } from "./agent_harness";
 *
 *   const add = defineTool({
 *     name: "add",
 *     description: "Add two numbers",
 *     parameters: {
 *       type: "object",
 *       properties: { a: { type: "number" }, b: { type: "number" } },
 *       required: ["a", "b"],
 *     },
 *     execute: async ({ a, b }) => a + b,
 *   });
 *
 *   const agent = new Agent({ model: "gpt-4o" });
 *   agent.registerTool(add);
 *   const result = await agent.run("What is 2 + 3?");
 */

import OpenAI from "openai";
import type {
  ChatCompletionMessageParam,
  ChatCompletionTool,
  ChatCompletionMessageToolCall,
} from "openai/resources/chat/completions";

// ---------------------------------------------------------------------------
// Hook events
// ---------------------------------------------------------------------------

export enum HookEvent {
  RunStart = "run_start",
  RunEnd = "run_end",
  LlmRequest = "llm_request",
  LlmResponse = "llm_response",
  ToolCall = "tool_call",
  ToolResult = "tool_result",
  ToolError = "tool_error",
  Retry = "retry",
  TokenStream = "token_stream",
  Error = "error",
}

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

export interface ToolDef<TArgs = any, TResult = any> {
  name: string;
  description: string;
  parameters: Record<string, any>; // JSON Schema
  execute: (args: TArgs) => TResult | Promise<TResult>;
}

export interface RunContext {
  agent: Agent;
  turn: number;
  metadata: Record<string, any>;
}

type HookCallback = (...args: any[]) => void | Promise<void>;

// ---------------------------------------------------------------------------
// Middleware
// ---------------------------------------------------------------------------

export interface Middleware {
  pre?(
    messages: ChatCompletionMessageParam[],
    context: RunContext
  ): ChatCompletionMessageParam[] | Promise<ChatCompletionMessageParam[]>;

  post?(
    message: Record<string, any>,
    context: RunContext
  ): Record<string, any> | Promise<Record<string, any>>;
}

// ---------------------------------------------------------------------------
// Tool definition helper
// ---------------------------------------------------------------------------

export function defineTool<TArgs = any, TResult = any>(
  def: ToolDef<TArgs, TResult>
): ToolDef<TArgs, TResult> {
  return def;
}

// ---------------------------------------------------------------------------
// Agent options
// ---------------------------------------------------------------------------

export interface AgentOptions {
  /** OpenAI-compatible model string, e.g. "gpt-4o", "claude-sonnet-4-20250514" */
  model?: string;
  /** System prompt */
  system?: string;
  /** Max tool-use loop iterations */
  maxTurns?: number;
  /** Retries on transient LLM errors */
  maxRetries?: number;
  /** Stream tokens and emit TokenStream hooks */
  stream?: boolean;
  /** Execute multiple tool calls concurrently */
  parallelToolCalls?: boolean;
  /**
   * Pre-configured OpenAI client. Pass this to point at a litellm proxy,
   * Azure, or any OpenAI-compatible endpoint.
   */
  client?: OpenAI;
  /** Extra params forwarded to chat.completions.create() */
  completionParams?: Partial<OpenAI.ChatCompletionCreateParams>;
}

// ---------------------------------------------------------------------------
// Agent
// ---------------------------------------------------------------------------

export class Agent {
  readonly model: string;
  readonly system: string | null;
  readonly maxTurns: number;
  readonly maxRetries: number;
  readonly stream: boolean;
  readonly parallelToolCalls: boolean;
  readonly completionParams: Partial<OpenAI.ChatCompletionCreateParams>;

  private client: OpenAI;
  private tools = new Map<string, ToolDef>();
  private middlewares: Middleware[] = [];
  private hooks = new Map<HookEvent, HookCallback[]>();

  constructor(opts: AgentOptions = {}) {
    this.model = opts.model ?? "gpt-4o";
    this.system = opts.system ?? null;
    this.maxTurns = opts.maxTurns ?? 20;
    this.maxRetries = opts.maxRetries ?? 2;
    this.stream = opts.stream ?? false;
    this.parallelToolCalls = opts.parallelToolCalls ?? true;
    this.completionParams = opts.completionParams ?? {};

    this.client =
      opts.client ??
      new OpenAI({
        // Falls back to OPENAI_API_KEY / OPENAI_BASE_URL env vars
      });

    for (const event of Object.values(HookEvent)) {
      this.hooks.set(event, []);
    }
  }

  // -- Registration --------------------------------------------------------

  registerTool(tool: ToolDef): void {
    this.tools.set(tool.name, tool);
  }

  use(mw: Middleware): void {
    this.middlewares.push(mw);
  }

  on(event: HookEvent, callback: HookCallback): void {
    this.hooks.get(event)!.push(callback);
  }

  // -- Schema generation ---------------------------------------------------

  private toolsSchema(): ChatCompletionTool[] {
    return Array.from(this.tools.values()).map((t) => ({
      type: "function" as const,
      function: {
        name: t.name,
        description: t.description,
        parameters: t.parameters,
      },
    }));
  }

  // -- Hook dispatch -------------------------------------------------------

  private async emit(event: HookEvent, ...args: any[]): Promise<void> {
    const cbs = this.hooks.get(event);
    if (!cbs?.length) return;
    const results = await Promise.allSettled(cbs.map((cb) => cb(...args)));
    for (const r of results) {
      if (r.status === "rejected") {
        console.warn(`Hook ${event} error:`, r.reason);
      }
    }
  }

  // -- Middleware pipeline --------------------------------------------------

  private async runPre(
    messages: ChatCompletionMessageParam[],
    ctx: RunContext
  ): Promise<ChatCompletionMessageParam[]> {
    for (const mw of this.middlewares) {
      if (mw.pre) {
        messages = await mw.pre(messages, ctx);
      }
    }
    return messages;
  }

  private async runPost(
    message: Record<string, any>,
    ctx: RunContext
  ): Promise<Record<string, any>> {
    for (const mw of this.middlewares) {
      if (mw.post) {
        message = await mw.post(message, ctx);
      }
    }
    return message;
  }

  // -- Tool execution ------------------------------------------------------

  private async executeTool(
    name: string,
    args: Record<string, any>
  ): Promise<string> {
    const tool = this.tools.get(name);
    if (!tool) {
      const err = `Unknown tool: ${name}`;
      await this.emit(HookEvent.ToolError, name, err);
      return JSON.stringify({ error: err });
    }

    await this.emit(HookEvent.ToolCall, name, args);
    try {
      const result = await tool.execute(args);
      await this.emit(HookEvent.ToolResult, name, result);
      return typeof result === "string" ? result : JSON.stringify(result);
    } catch (err: any) {
      await this.emit(HookEvent.ToolError, name, err);
      return JSON.stringify({ error: err?.message ?? String(err) });
    }
  }

  // -- LLM call ------------------------------------------------------------

  private async callLlm(
    messages: ChatCompletionMessageParam[],
    ctx: RunContext
  ): Promise<Record<string, any>> {
    const toolsDef = this.tools.size > 0 ? this.toolsSchema() : undefined;
    await this.emit(HookEvent.LlmRequest, messages, toolsDef);

    let lastErr: Error | null = null;

    for (let attempt = 1; attempt <= this.maxRetries + 1; attempt++) {
      try {
        if (this.stream) {
          return await this.callLlmStream(messages, toolsDef, ctx);
        }

        const resp = await this.client.chat.completions.create({
          model: this.model,
          messages,
          tools: toolsDef,
          ...this.completionParams,
        } as OpenAI.ChatCompletionCreateParams);

        const msg = resp.choices[0].message as unknown as Record<string, any>;
        await this.emit(HookEvent.LlmResponse, msg);
        return msg;
      } catch (err: any) {
        lastErr = err;
        if (attempt <= this.maxRetries) {
          await this.emit(HookEvent.Retry, attempt, err);
          await sleep(Math.min(2 ** attempt * 1000, 10_000));
        }
      }
    }

    throw new Error(
      `LLM call failed after ${this.maxRetries + 1} attempts: ${lastErr?.message}`
    );
  }

  private async callLlmStream(
    messages: ChatCompletionMessageParam[],
    toolsDef: ChatCompletionTool[] | undefined,
    _ctx: RunContext
  ): Promise<Record<string, any>> {
    const stream = await this.client.chat.completions.create({
      model: this.model,
      messages,
      tools: toolsDef,
      stream: true,
      ...this.completionParams,
    } as OpenAI.ChatCompletionCreateParams & { stream: true });

    const contentParts: string[] = [];
    const toolCallsAccum = new Map<
      number,
      { id: string; type: "function"; function: { name: string; arguments: string } }
    >();

    for await (const chunk of stream) {
      const delta = chunk.choices[0]?.delta;
      if (!delta) continue;

      if (delta.content) {
        contentParts.push(delta.content);
        await this.emit(HookEvent.TokenStream, delta.content);
      }

      if (delta.tool_calls) {
        for (const tc of delta.tool_calls) {
          const idx = tc.index;
          if (!toolCallsAccum.has(idx)) {
            toolCallsAccum.set(idx, {
              id: tc.id ?? "",
              type: "function",
              function: { name: "", arguments: "" },
            });
          }
          const entry = toolCallsAccum.get(idx)!;
          if (tc.id) entry.id = tc.id;
          if (tc.function?.name) entry.function.name += tc.function.name;
          if (tc.function?.arguments)
            entry.function.arguments += tc.function.arguments;
        }
      }
    }

    const msg: Record<string, any> = {
      role: "assistant",
      content: contentParts.join("") || null,
    };

    if (toolCallsAccum.size > 0) {
      msg.tool_calls = Array.from(toolCallsAccum.entries())
        .sort(([a], [b]) => a - b)
        .map(([, v]) => v);
    }

    await this.emit(HookEvent.LlmResponse, msg);
    return msg;
  }

  // -- Main loop -----------------------------------------------------------

  async run(
    prompt: string,
    options?: {
      messages?: ChatCompletionMessageParam[];
      contextMeta?: Record<string, any>;
    }
  ): Promise<string> {
    const ctx: RunContext = {
      agent: this,
      turn: 0,
      metadata: options?.contextMeta ?? {},
    };

    const msgs: ChatCompletionMessageParam[] = [];
    if (this.system) {
      msgs.push({ role: "system", content: this.system });
    }
    if (options?.messages) {
      msgs.push(...structuredClone(options.messages));
    }
    msgs.push({ role: "user", content: prompt });

    await this.emit(HookEvent.RunStart, msgs);

    for (let turn = 0; turn < this.maxTurns; turn++) {
      ctx.turn = turn;

      // Middleware pre
      const callMsgs = await this.runPre(structuredClone(msgs), ctx);

      // LLM call
      let assistantMsg: Record<string, any>;
      try {
        assistantMsg = await this.callLlm(
          callMsgs as ChatCompletionMessageParam[],
          ctx
        );
      } catch (err: any) {
        await this.emit(HookEvent.Error, err);
        throw err;
      }

      // Middleware post
      assistantMsg = await this.runPost(assistantMsg, ctx);
      msgs.push(assistantMsg as ChatCompletionMessageParam);

      // No tool calls → done
      const toolCalls = assistantMsg.tool_calls as
        | ChatCompletionMessageToolCall[]
        | undefined;

      if (!toolCalls?.length) {
        const final = (assistantMsg.content as string) ?? "";
        await this.emit(HookEvent.RunEnd, msgs, final);
        return final;
      }

      // Execute tool calls
      if (this.parallelToolCalls && toolCalls.length > 1) {
        const results = await Promise.all(
          toolCalls.map((tc) =>
            this.executeTool(
              tc.function.name,
              safeJsonParse(tc.function.arguments)
            )
          )
        );
        for (let i = 0; i < toolCalls.length; i++) {
          msgs.push({
            role: "tool",
            tool_call_id: toolCalls[i].id,
            content: results[i],
          });
        }
      } else {
        for (const tc of toolCalls) {
          const result = await this.executeTool(
            tc.function.name,
            safeJsonParse(tc.function.arguments)
          );
          msgs.push({
            role: "tool",
            tool_call_id: tc.id,
            content: result,
          });
        }
      }
    }

    throw new Error(`Agent exceeded maxTurns (${this.maxTurns})`);
  }
}

// ---------------------------------------------------------------------------
// Utilities
// ---------------------------------------------------------------------------

function safeJsonParse(s: string): Record<string, any> {
  try {
    return JSON.parse(s);
  } catch {
    return {};
  }
}

function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

// ---------------------------------------------------------------------------
// Built-in middleware
// ---------------------------------------------------------------------------

export class TruncationMiddleware implements Middleware {
  constructor(private maxMessages = 40) {}

  async pre(
    messages: ChatCompletionMessageParam[],
    _ctx: RunContext
  ): Promise<ChatCompletionMessageParam[]> {
    if (messages.length <= this.maxMessages) return messages;
    const system = messages.filter((m) => m.role === "system");
    const rest = messages.filter((m) => m.role !== "system");
    return [...system, ...rest.slice(-(this.maxMessages - system.length))];
  }
}

export class LoggingMiddleware implements Middleware {
  async pre(
    messages: ChatCompletionMessageParam[],
    ctx: RunContext
  ): Promise<ChatCompletionMessageParam[]> {
    console.debug(`[turn ${ctx.turn}] Sending ${messages.length} messages`);
    return messages;
  }

  async post(
    message: Record<string, any>,
    ctx: RunContext
  ): Promise<Record<string, any>> {
    const tc = message.tool_calls;
    if (tc) {
      const names = tc.map((t: any) => t.function.name);
      console.debug(`[turn ${ctx.turn}] Tool calls: ${names.join(", ")}`);
    } else {
      const snippet = ((message.content as string) ?? "").slice(0, 120);
      console.debug(`[turn ${ctx.turn}] Response: ${snippet}...`);
    }
    return message;
  }
}

// ---------------------------------------------------------------------------
// Demo
// ---------------------------------------------------------------------------

async function main() {
  const add = defineTool({
    name: "add",
    description: "Add two numbers together",
    parameters: {
      type: "object",
      properties: {
        a: { type: "number", description: "First number" },
        b: { type: "number", description: "Second number" },
      },
      required: ["a", "b"],
    },
    execute: async ({ a, b }: { a: number; b: number }) => a + b,
  });

  const getWeather = defineTool({
    name: "get_weather",
    description: "Get the current weather for a city (stub)",
    parameters: {
      type: "object",
      properties: {
        city: { type: "string", description: "City name" },
      },
      required: ["city"],
    },
    execute: async ({ city }: { city: string }) => ({
      city,
      temp_f: 72,
      condition: "sunny",
    }),
  });

  const agent = new Agent({
    model: process.env.AGENT_MODEL ?? "gpt-4o",
    system: "You are a helpful assistant. Use your tools when appropriate.",
    maxTurns: 10,
    parallelToolCalls: true,
    completionParams: { temperature: 0.3 },
  });

  agent.registerTool(add);
  agent.registerTool(getWeather);

  agent.use(new LoggingMiddleware());
  agent.use(new TruncationMiddleware(30));

  agent.on(HookEvent.ToolCall, (name, args) =>
    console.log(`  🔧 ${name}(${JSON.stringify(args)})`)
  );
  agent.on(HookEvent.ToolResult, (name, res) =>
    console.log(`  ✅ ${name} → ${JSON.stringify(res)}`)
  );
  agent.on(HookEvent.TokenStream, (tok) => process.stdout.write(tok));

  const result = await agent.run(
    "What's the weather in Tokyo, and what's 17 + 38?"
  );
  console.log("\n---\n", result);
}

// Run if executed directly
main().catch(console.error);

File: examples/agent_harness.py
"""
agent_harness.py — A lightweight, async-native, single-file LLM agent harness.

Built on litellm for multi-provider support. Features:
  - Decorator-based tool registration with auto-schema generation
  - Async middleware pipeline (pre/post processing of messages and responses)
  - Async hook system for lifecycle events
  - Streaming support
  - Parallel tool execution
  - Configurable retry with exponential backoff

Usage:
    import asyncio
    from agent_harness import Agent, tool

    @tool(description="Add two numbers")
    async def add(a: int, b: int) -> int:
        return a + b

    async def main():
        agent = Agent(model="anthropic/claude-sonnet-4-20250514")
        agent.register_tool(add)
        result = await agent.run("What is 2 + 3?")
        print(result)

    asyncio.run(main())
"""

from __future__ import annotations

import asyncio
import inspect
import json
import logging
import traceback
from copy import deepcopy
from dataclasses import dataclass, field
from enum import Enum
from typing import (
    Any,
    Callable,
    Protocol,
    TypeVar,
    get_type_hints,
    runtime_checkable,
)

import litellm

# ---------------------------------------------------------------------------
# Logging
# ---------------------------------------------------------------------------
log = logging.getLogger("agent_harness")

# ---------------------------------------------------------------------------
# Type helpers
# ---------------------------------------------------------------------------
PYTHON_TYPE_TO_JSON: dict[type, str] = {
    str: "string",
    int: "integer",
    float: "number",
    bool: "boolean",
    list: "array",
    dict: "object",
}

F = TypeVar("F", bound=Callable[..., Any])


# ---------------------------------------------------------------------------
# Hook events
# ---------------------------------------------------------------------------
class HookEvent(str, Enum):
    """Lifecycle events the agent emits."""

    RUN_START = "run_start"          # (messages,)
    RUN_END = "run_end"              # (messages, final_text)
    LLM_REQUEST = "llm_request"      # (messages, tools_schema)
    LLM_RESPONSE = "llm_response"    # (response,)
    TOOL_CALL = "tool_call"          # (tool_name, arguments)
    TOOL_RESULT = "tool_result"      # (tool_name, result)
    TOOL_ERROR = "tool_error"        # (tool_name, error)
    RETRY = "retry"                  # (attempt, error)
    TOKEN_STREAM = "token_stream"    # (token_str,)
    ERROR = "error"                  # (error,)


# ---------------------------------------------------------------------------
# Middleware protocol
# ---------------------------------------------------------------------------
@runtime_checkable
class Middleware(Protocol):
    """
    Middleware contract. Implement any subset of these async methods.
    - pre:  transform the messages list before each LLM call
    - post: transform the assistant message after each LLM call
    """

    async def pre(self, messages: list[dict], context: RunContext) -> list[dict]: ...
    async def post(self, message: dict, context: RunContext) -> dict: ...


class BaseMiddleware:
    """Convenience base — override only what you need."""

    async def pre(self, messages: list[dict], context: RunContext) -> list[dict]:
        return messages

    async def post(self, message: dict, context: RunContext) -> dict:
        return message


# ---------------------------------------------------------------------------
# Run context (passed through middleware and hooks)
# ---------------------------------------------------------------------------
@dataclass
class RunContext:
    """Mutable bag of state for the current run, accessible to middleware & hooks."""

    agent: Agent
    turn: int = 0
    metadata: dict[str, Any] = field(default_factory=dict)


# ---------------------------------------------------------------------------
# Tool registry helpers
# ---------------------------------------------------------------------------
@dataclass
class ToolDef:
    """Internal representation of a registered tool."""

    name: str
    description: str
    function: Callable[..., Any]  # sync or async
    parameters: dict[str, Any]   # JSON Schema for the parameters


def _build_param_schema(fn: Callable) -> dict[str, Any]:
    """Derive a JSON Schema 'parameters' object from a function's type hints."""
    hints = get_type_hints(fn)
    sig = inspect.signature(fn)
    properties: dict[str, Any] = {}
    required: list[str] = []

    for name, param in sig.parameters.items():
        json_type = PYTHON_TYPE_TO_JSON.get(hints.get(name, str), "string")
        prop: dict[str, Any] = {"type": json_type}

        if param.default is inspect.Parameter.empty:
            required.append(name)
        else:
            prop["default"] = param.default

        properties[name] = prop

    return {
        "type": "object",
        "properties": properties,
        "required": required,
    }


def tool(
    description: str = "",
    name: str | None = None,
    schema_override: dict | None = None,
) -> Callable[[F], F]:
    """
    Decorator that marks a function as an agent tool.
    Works with both sync and async functions.

        @tool(description="Search the web")
        async def web_search(query: str) -> str: ...

    The decorated function gains a `_tool_def` attribute used during registration.
    You can pass `schema_override` to supply your own JSON Schema for parameters.
    """

    def decorator(fn: F) -> F:
        fn._tool_def = ToolDef(  # type: ignore[attr-defined]
            name=name or fn.__name__,
            description=description or fn.__doc__ or "",
            function=fn,
            parameters=schema_override or _build_param_schema(fn),
        )
        return fn

    return decorator


# ---------------------------------------------------------------------------
# Helpers
# ---------------------------------------------------------------------------

async def _call_fn(fn: Callable, *args: Any, **kwargs: Any) -> Any:
    """Call a sync or async function uniformly, always returning an awaitable."""
    result = fn(*args, **kwargs)
    if inspect.isawaitable(result):
        return await result
    return result


def _safe_json_loads(s: str) -> dict:
    try:
        return json.loads(s)
    except (json.JSONDecodeError, TypeError):
        return {}


# ---------------------------------------------------------------------------
# Agent
# ---------------------------------------------------------------------------
class Agent:
    """
    Lightweight async agent loop.

    Parameters
    ----------
    model : str
        A litellm model string, e.g. "anthropic/claude-sonnet-4-20250514",
        "openai/gpt-4o", "ollama/llama3", etc.
    system : str | None
        System prompt.
    max_turns : int
        Safety cap on tool-use loops per run.
    max_retries : int
        Retries on transient LLM errors.
    stream : bool
        If True, stream tokens and emit TOKEN_STREAM hooks.
    parallel_tool_calls : bool
        If True, execute multiple tool calls concurrently via asyncio.gather.
    litellm_kwargs : dict
        Extra kwargs forwarded to litellm.acompletion() (temperature, etc.)
    """

    def __init__(
        self,
        model: str = "anthropic/claude-sonnet-4-20250514",
        system: str | None = None,
        max_turns: int = 20,
        max_retries: int = 2,
        stream: bool = False,
        parallel_tool_calls: bool = True,
        **litellm_kwargs: Any,
    ):
        self.model = model
        self.system = system
        self.max_turns = max_turns
        self.max_retries = max_retries
        self.stream = stream
        self.parallel_tool_calls = parallel_tool_calls
        self.litellm_kwargs = litellm_kwargs

        self._tools: dict[str, ToolDef] = {}
        self._middleware: list[BaseMiddleware | Middleware] = []
        self._hooks: dict[HookEvent, list[Callable]] = {e: [] for e in HookEvent}

    # -- Registration -------------------------------------------------------

    def register_tool(self, fn_or_def: Callable | ToolDef) -> None:
        """Register a @tool-decorated function or a raw ToolDef."""
        if isinstance(fn_or_def, ToolDef):
            td = fn_or_def
        elif hasattr(fn_or_def, "_tool_def"):
            td = fn_or_def._tool_def  # type: ignore[attr-defined]
        else:
            raise ValueError(
                "Pass a @tool-decorated function or a ToolDef instance."
            )
        self._tools[td.name] = td
        log.info("Registered tool: %s", td.name)

    def use(self, mw: BaseMiddleware | Middleware) -> None:
        """Add middleware to the pipeline (order matters — FIFO)."""
        self._middleware.append(mw)

    def on(self, event: HookEvent, callback: Callable) -> None:
        """Subscribe to a lifecycle event. Callback can be sync or async."""
        self._hooks[event].append(callback)

    # -- Schema generation --------------------------------------------------

    def _tools_schema(self) -> list[dict]:
        """Build the litellm/OpenAI-compatible tools array."""
        return [
            {
                "type": "function",
                "function": {
                    "name": td.name,
                    "description": td.description,
                    "parameters": td.parameters,
                },
            }
            for td in self._tools.values()
        ]

    # -- Hook dispatch ------------------------------------------------------

    async def _emit(self, event: HookEvent, *args: Any) -> None:
        """Fire all hooks for an event concurrently."""
        if not self._hooks[event]:
            return
        results = await asyncio.gather(
            *(_call_fn(cb, *args) for cb in self._hooks[event]),
            return_exceptions=True,
        )
        for r in results:
            if isinstance(r, Exception):
                log.warning("Hook %s raised: %s", event, r)

    # -- Middleware pipeline -------------------------------------------------

    async def _run_pre(self, messages: list[dict], ctx: RunContext) -> list[dict]:
        for mw in self._middleware:
            if hasattr(mw, "pre"):
                messages = await _call_fn(mw.pre, messages, ctx)
        return messages

    async def _run_post(self, message: dict, ctx: RunContext) -> dict:
        for mw in self._middleware:
            if hasattr(mw, "post"):
                message = await _call_fn(mw.post, message, ctx)
        return message

    # -- Tool execution -----------------------------------------------------

    async def _execute_tool(self, name: str, arguments: dict) -> str:
        """Run a tool and return a JSON-serialised result string."""
        td = self._tools.get(name)
        if td is None:
            err = f"Unknown tool: {name}"
            await self._emit(HookEvent.TOOL_ERROR, name, err)
            return json.dumps({"error": err})

        await self._emit(HookEvent.TOOL_CALL, name, arguments)
        try:
            result = await _call_fn(td.function, **arguments)
            await self._emit(HookEvent.TOOL_RESULT, name, result)
            return result if isinstance(result, str) else json.dumps(result)
        except Exception as exc:
            await self._emit(HookEvent.TOOL_ERROR, name, exc)
            return json.dumps({"error": str(exc)})

    # -- LLM call (with retries + optional streaming) -----------------------

    async def _call_llm(self, messages: list[dict], ctx: RunContext) -> dict:
        """Make one LLM call, returning the assistant message dict."""
        tools = self._tools_schema() or None
        await self._emit(HookEvent.LLM_REQUEST, messages, tools)

        last_err: Exception | None = None
        for attempt in range(1, self.max_retries + 2):
            try:
                if self.stream:
                    return await self._call_llm_stream(messages, tools, ctx)

                resp = await litellm.acompletion(
                    model=self.model,
                    messages=messages,
                    tools=tools,
                    **self.litellm_kwargs,
                )
                msg = resp.choices[0].message.model_dump()  # type: ignore[union-attr]
                await self._emit(HookEvent.LLM_RESPONSE, msg)
                return msg

            except Exception as exc:
                last_err = exc
                if attempt <= self.max_retries:
                    await self._emit(HookEvent.RETRY, attempt, exc)
                    log.warning("Retry %d/%d: %s", attempt, self.max_retries, exc)
                    await asyncio.sleep(min(2**attempt, 10))

        raise RuntimeError(
            f"LLM call failed after {self.max_retries + 1} attempts: {last_err}"
        )

    async def _call_llm_stream(
        self, messages: list[dict], tools: list | None, ctx: RunContext
    ) -> dict:
        """Stream tokens, accumulate, and return the full message dict."""
        resp = await litellm.acompletion(
            model=self.model,
            messages=messages,
            tools=tools,
            stream=True,
            **self.litellm_kwargs,
        )

        content_parts: list[str] = []
        tool_calls_accum: dict[int, dict] = {}

        async for chunk in resp:
            delta = chunk.choices[0].delta  # type: ignore[union-attr]

            if delta.content:
                content_parts.append(delta.content)
                await self._emit(HookEvent.TOKEN_STREAM, delta.content)

            if delta.tool_calls:
                for tc in delta.tool_calls:
                    idx = tc.index
                    if idx not in tool_calls_accum:
                        tool_calls_accum[idx] = {
                            "id": tc.id or "",
                            "type": "function",
                            "function": {"name": "", "arguments": ""},
                        }
                    entry = tool_calls_accum[idx]
                    if tc.id:
                        entry["id"] = tc.id
                    if tc.function:
                        if tc.function.name:
                            entry["function"]["name"] += tc.function.name
                        if tc.function.arguments:
                            entry["function"]["arguments"] += tc.function.arguments

        msg: dict[str, Any] = {
            "role": "assistant",
            "content": "".join(content_parts) or None,
        }
        if tool_calls_accum:
            msg["tool_calls"] = [tool_calls_accum[i] for i in sorted(tool_calls_accum)]

        await self._emit(HookEvent.LLM_RESPONSE, msg)
        return msg

    # -- Main loop ----------------------------------------------------------

    async def run(
        self,
        prompt: str,
        messages: list[dict] | None = None,
        context_meta: dict[str, Any] | None = None,
    ) -> str:
        """
        Run the agent to completion and return the final text response.

        Parameters
        ----------
        prompt : str
            The user message.
        messages : list[dict] | None
            Optional conversation history to continue from.
        context_meta : dict | None
            Arbitrary metadata attached to the RunContext for middleware/hooks.
        """
        ctx = RunContext(agent=self, metadata=context_meta or {})

        msgs: list[dict] = []
        if self.system:
            msgs.append({"role": "system", "content": self.system})
        if messages:
            msgs.extend(deepcopy(messages))
        msgs.append({"role": "user", "content": prompt})

        await self._emit(HookEvent.RUN_START, msgs)

        for turn in range(self.max_turns):
            ctx.turn = turn

            # Middleware pre-processing
            call_msgs = await self._run_pre(deepcopy(msgs), ctx)

            # LLM call
            try:
                assistant_msg = await self._call_llm(call_msgs, ctx)
            except Exception as exc:
                await self._emit(HookEvent.ERROR, exc)
                raise

            # Middleware post-processing
            assistant_msg = await self._run_post(assistant_msg, ctx)
            msgs.append(assistant_msg)

            # If no tool calls, we're done
            tool_calls = assistant_msg.get("tool_calls")
            if not tool_calls:
                final = assistant_msg.get("content", "") or ""
                await self._emit(HookEvent.RUN_END, msgs, final)
                return final

            # Execute tool calls — parallel or sequential
            if self.parallel_tool_calls and len(tool_calls) > 1:
                results = await asyncio.gather(
                    *(
                        self._execute_tool(
                            tc["function"]["name"],
                            _safe_json_loads(tc["function"]["arguments"]),
                        )
                        for tc in tool_calls
                    )
                )
                for tc, result_str in zip(tool_calls, results):
                    msgs.append({
                        "role": "tool",
                        "tool_call_id": tc["id"],
                        "content": result_str,
                    })
            else:
                for tc in tool_calls:
                    result_str = await self._execute_tool(
                        tc["function"]["name"],
                        _safe_json_loads(tc["function"]["arguments"]),
                    )
                    msgs.append({
                        "role": "tool",
                        "tool_call_id": tc["id"],
                        "content": result_str,
                    })

        raise RuntimeError(f"Agent exceeded max_turns ({self.max_turns})")


# ===========================================================================
# Built-in middleware (import and .use() as needed)
# ===========================================================================


class TruncationMiddleware(BaseMiddleware):
    """Keep the message list under a budget by dropping old turns."""

    def __init__(self, max_messages: int = 40):
        self.max_messages = max_messages

    async def pre(self, messages: list[dict], context: RunContext) -> list[dict]:
        if len(messages) <= self.max_messages:
            return messages
        system = [m for m in messages if m["role"] == "system"]
        rest = [m for m in messages if m["role"] != "system"]
        return system + rest[-(self.max_messages - len(system)) :]


class LoggingMiddleware(BaseMiddleware):
    """Log every LLM request/response for debugging."""

    async def pre(self, messages: list[dict], context: RunContext) -> list[dict]:
        log.debug("[turn %d] Sending %d messages", context.turn, len(messages))
        return messages

    async def post(self, message: dict, context: RunContext) -> dict:
        tc = message.get("tool_calls")
        if tc:
            names = [t["function"]["name"] for t in tc]
            log.debug("[turn %d] Tool calls: %s", context.turn, names)
        else:
            snippet = (message.get("content") or "")[:120]
            log.debug("[turn %d] Response: %s...", context.turn, snippet)
        return message


class CostTrackingMiddleware(BaseMiddleware):
    """Accumulate token usage in context.metadata["usage"]."""

    async def post(self, message: dict, context: RunContext) -> dict:
        usage = context.metadata.setdefault(
            "usage", {"prompt_tokens": 0, "completion_tokens": 0}
        )
        if "_usage" in message:
            u = message["_usage"]
            usage["prompt_tokens"] += u.get("prompt_tokens", 0)
            usage["completion_tokens"] += u.get("completion_tokens", 0)
        return message


# ===========================================================================
# Demo
# ===========================================================================
if __name__ == "__main__":
    import os

    logging.basicConfig(level=logging.DEBUG)

    # --- Tools (async and sync both work) ---

    @tool(description="Add two numbers together")
    async def add(a: int, b: int) -> int:
        return a + b

    @tool(description="Get the current weather for a city (stub)")
    async def get_weather(city: str) -> dict:
        await asyncio.sleep(0.1)  # simulate latency
        return {"city": city, "temp_f": 72, "condition": "sunny"}

    @tool(description="Multiply two numbers")
    def multiply(a: int, b: int) -> int:
        """Plain sync tool — also works."""
        return a * b

    # --- Custom middleware ---

    class InjectTimestamp(BaseMiddleware):
        async def pre(self, messages, context):
            from datetime import datetime, timezone

            ts = datetime.now(timezone.utc).isoformat()
            for m in messages:
                if m["role"] == "system":
                    m["content"] += f"\n\nCurrent time: {ts}"
                    break
            return messages

    # --- Agent setup ---

    agent = Agent(
        model=os.getenv("AGENT_MODEL", "anthropic/claude-sonnet-4-20250514"),
        system="You are a helpful assistant. Use your tools when appropriate.",
        max_turns=10,
        stream=False,
        parallel_tool_calls=True,
        temperature=0.3,
    )

    agent.register_tool(add)
    agent.register_tool(get_weather)
    agent.register_tool(multiply)

    agent.use(LoggingMiddleware())
    agent.use(TruncationMiddleware(max_messages=30))
    agent.use(InjectTimestamp())

    # Hooks — sync and async both work
    agent.on(HookEvent.TOOL_CALL, lambda name, args: print(f"  🔧 {name}({args})"))
    agent.on(HookEvent.TOOL_RESULT, lambda name, res: print(f"  ✅ {name} → {res}"))

    async def on_stream(tok: str) -> None:
        print(tok, end="", flush=True)

    agent.on(HookEvent.TOKEN_STREAM, on_stream)

    async def main() -> None:
        result = await agent.run(
            "What's the weather in Tokyo, what's 17 + 38, and what's 6 * 7?"
        )
        print("\n---\n", result)

    asyncio.run(main())
